### DISCUSSION TITLE
property overflow = Hidden leads to cut-off decimals with DOCX output

### DISCUSSION THREAD

**Discussion by reiner-killet** [eclipse-birt/birt#916] [General]
Hi,
since BIRT 4.8 we watch decimals being cut-off at the end of float values in table cells by the BIRT runtime (not RCP designer) with DOCX output.
![grafik](https://user-images.githubusercontent.com/95427074/161710315-83fbe26f-cc6d-42ae-8516-471ff5378efc.png)
I first tried to set "Overflow" to "Visible" for the entire table. This had no effect. Next, I examined inheritance of "Overflow" for rows, columns, cells, and data tags. I saw that it is not inherited from outside to inside. All inner elements were still showing "Overflow = Hidden (inherited)" what is wrong. Last, I set "Overflow" to "Visible" for the data tags. This works. Questions:
1. Can you please check if this is a bug which came to existence somewhere between version 4.6 and 4.8.?
2. Can someone explain why "Overflow" is not inherited?
3. Is there a simple way to update a large number of templates to achieve a behaviour like in 4.6?
Thanks a lot

**Comment by hvbtup** [comment]
@reiner-killet Can you create a simple report demonstrating the issue?

**Reply by reiner-killet** [reply]
<?xml version="1.0" encoding="UTF-8"?>
<report xmlns="http://www.eclipse.org/birt/2005/design" version="3.2.23" id="1">
    <property name="createdBy">Eclipse BIRT Designer Version 4.8.0.v201806261756</property>
    <property name="units">in</property>
    <property name="iconFile">/templates/blank_report.gif</property>
    <property name="bidiLayoutOrientation">ltr</property>
    <property name="imageDPI">120</property>
    <data-sources>
        <oda-data-source extensionID="org.eclipse.datatools.enablement.oda.xml" name="Data Source" id="5">
            <property name="FILELIST">C:\temp\birt\Test\CSO_ART_PRI_TST_GER.xml</property>
        </oda-data-source>
    </data-sources>
    <data-sets>
        <oda-data-set extensionID="org.eclipse.datatools.enablement.oda.xml.dataSet" name="PRODUCTS" id="6">
            <list-property name="columnHints">
                <structure>
                    <property name="columnName">NAME</property>
                    <property name="analysis">dimension</property>
                    <text-property name="heading">NAME</text-property>
                </structure>
                <structure>
                    <property name="columnName">PRICE</property>
                    <property name="analysis">dimension</property>
                    <text-property name="heading">PRICE</text-property>
                </structure>
            </list-property>
            <list-property name="parameters"/>
            <structure name="cachedMetaData">
                <list-property name="resultSet">
                    <structure>
                        <property name="position">1</property>
                        <property name="name">NAME</property>
                        <property name="dataType">string</property>
                    </structure>
                    <structure>
                        <property name="position">2</property>
                        <property name="name">PRICE</property>
                        <property name="dataType">float</property>
                    </structure>
                </list-property>
            </structure>
            <property name="dataSource">Data Source</property>
            <list-property name="resultSet">
                <structure>
                    <property name="position">1</property>
                    <property name="name">NAME</property>
                    <property name="nativeName">NAME</property>
                    <property name="dataType">string</property>
                    <property name="nativeDataType">12</property>
                </structure>
                <structure>
                    <property name="position">2</property>
                    <property name="name">PRICE</property>
                    <property name="nativeName">PRICE</property>
                    <property name="dataType">float</property>
                    <property name="nativeDataType">8</property>
                </structure>
            </list-property>
            <xml-property name="queryText"><![CDATA[table0#-TNAME-#table0#:#[/TOP/PRODUCT/DAT]#:#{NAME;STRING;/NAME},{PRICE;DOUBLE;/PRICE}]]></xml-property>
            <xml-property name="designerValues"><![CDATA[<?xml version="1.0" encoding="UTF-8"?>
<model:DesignValues xmlns:design="http://www.eclipse.org/datatools/connectivity/oda/design" xmlns:model="http://www.eclipse.org/birt/report/model/adapter/odaModel">
  <Version>2.0</Version>
  <design:ResultSets derivedMetaData="true">
    <design:resultSetDefinitions>
      <design:resultSetColumns>
        <design:resultColumnDefinitions>
          <design:attributes>
            <design:identifier>
              <design:name>NAME</design:name>
              <design:position>1</design:position>
            </design:identifier>
            <design:nativeDataTypeCode>12</design:nativeDataTypeCode>
            <design:precision>-1</design:precision>
            <design:scale>-1</design:scale>
            <design:nullability>Unknown</design:nullability>
          </design:attributes>
          <design:usageHints>
            <design:label>NAME</design:label>
            <design:formattingHints/>
          </design:usageHints>
        </design:resultColumnDefinitions>
        <design:resultColumnDefinitions>
          <design:attributes>
            <design:identifier>
              <design:name>PRICE</design:name>
              <design:position>2</design:position>
            </design:identifier>
            <design:nativeDataTypeCode>12</design:nativeDataTypeCode>
            <design:precision>-1</design:precision>
            <design:scale>-1</design:scale>
            <design:nullability>Unknown</design:nullability>
          </design:attributes>
          <design:usageHints>
            <design:label>PRICE</design:label>
            <design:formattingHints/>
          </design:usageHints>
        </design:resultColumnDefinitions>
      </design:resultSetColumns>
      <design:criteria/>
    </design:resultSetDefinitions>
  </design:ResultSets>
</model:DesignValues>]]></xml-property>
            <list-property name="privateDriverProperties">
                <ex-property>
                    <name>MAX_ROW</name>
                    <value>-1</value>
                </ex-property>
                <ex-property>
                    <name>XML_FILE</name>
                </ex-property>
            </list-property>
        </oda-data-set>
    </data-sets>
    <page-setup>
        <simple-master-page name="Simple MasterPage" id="2">
            <property name="type">a4</property>
            <property name="orientation">portrait</property>
            <property name="topMargin">2cm</property>
            <property name="leftMargin">2cm</property>
            <property name="bottomMargin">2cm</property>
            <property name="rightMargin">2cm</property>
            <property name="headerHeight">1cm</property>
            <property name="footerHeight">1cm</property>
            <page-header>
                <label id="4">
                    <property name="fontFamily">sans-serif</property>
                    <property name="fontSize">12pt</property>
                    <text-property name="text">CUT-OFF DECIMALS SAMPLE</text-property>
                </label>
            </page-header>
            <page-footer>
                <text id="3">
                    <property name="contentType">html</property>
                    <text-property name="content"><![CDATA[<value-of>new Date()</value-of>]]></text-property>
                </text>
            </page-footer>
        </simple-master-page>
    </page-setup>
    <body>
        <table id="7">
            <property name="fontFamily">"Arial"</property>
            <property name="marginBottom">40pt</property>
            <property name="overflow">visible</property>
            <property name="dataSet">PRODUCTS</property>
            <list-property name="boundDataColumns">
                <structure>
                    <property name="name">NAME</property>
                    <text-property name="displayName">NAME</text-property>
                    <expression name="expression" type="javascript">dataSetRow["NAME"]</expression>
                    <property name="dataType">string</property>
                </structure>
                <structure>
                    <property name="name">PRICE</property>
                    <text-property name="displayName">PRICE</text-property>
                    <expression name="expression" type="javascript">dataSetRow["PRICE"]</expression>
                    <property name="dataType">float</property>
                    <property name="allowExport">true</property>
                </structure>
            </list-property>
            <column id="17">
                <property name="textAlign">left</property>
            </column>
            <column id="18">
                <property name="textAlign">right</property>
                <property name="width">2.2cm</property>
            </column>
            <header>
                <row id="8">
                    <property name="fontWeight">bold</property>
                    <cell id="9">
                        <label id="19">
                            <property name="overflow">auto</property>
                            <text-property name="text">Name</text-property>
                        </label>
                    </cell>
                    <cell id="10">
                        <label id="20">
                            <property name="overflow">auto</property>
                            <text-property name="text">Preisangabe Material</text-property>
                        </label>
                    </cell>
                </row>
            </header>
            <detail>
                <row id="11">
                    <cell id="12">
                        <data id="21">
                            <property name="overflow">auto</property>
                            <property name="resultSetColumn">NAME</property>
                        </data>
                    </cell>
                    <cell id="13">
                        <property name="whiteSpace">normal</property>
                        <property name="overflow">auto</property>
                        <data id="22">
                            <structure name="numberFormat">
                                <property name="category">Fixed</property>
                                <property name="pattern">#,##0.00{RoundingMode=HALF_UP}</property>
                            </structure>
                            <property name="whiteSpace">normal</property>
                            <property name="resultSetColumn">PRICE</property>
                        </data>
                    </cell>
                </row>
            </detail>
            <footer>
                <row id="14">
                    <cell id="15"/>
                    <cell id="16"/>
                </row>
            </footer>
        </table>
    </body>
</report>

**Reply by hvbtup** [reply]
Please attach the rptdesign in a *.zip file.

**Comment by hvbtup** [comment]
@reiner-killet Did you also test with BIRT 4.9?

**Reply by reiner-killet** [reply]
Yes. 4.9 release and 4.8 show same behaviour.

**Comment by hvbtup** [comment]
Regarding your second question:
If a property is inheritable or not is defined in the ROM specification(file `rom.def` in the sources).
The "overflow" property is defined to not be inherited and this never has changed AFAIK:

    <Property canInherit="false" detailType="overflow" displayNameID="Element.Style.overflow" isStyleProperty="true" name="overflow" since="2.5.2" type="choice">
        <Default>hidden</Default>
    </Property>

**Reply by reiner-killet** [reply]
OK, agree. But in this case the designer should not show "Inherited". 
![grafik](https://user-images.githubusercontent.com/95427074/161746828-dd0fa1aa-188a-4e34-9d2f-17160be82a20.png)

**Reply by hvbtup** [reply]
It is inherited from the default style even though it cannot be inherited from parent layout elements.

**Comment by hvbtup** [comment]
I digged a bit into the sources.
The handling of overflow: hidden was introduced with 8c4f519544b443fedc3b2d89fdbd8020a7dc320d in 2017 with the comment:

> 
> Overflow hidden property does not exist on table cell in DOCX
> 
> This enhancement is to crop the words that exceed the width of the cell.
> If the property is not set, it will wrap around cell and expand the
> height of the row and table changing the overall layout
> 

I think that the commit makes sense (though our word reports may be effected just like like yours and we'll probably have to test and change them).

But I can't reproduce the unintended cropping. For example, I took the sample report from  #875.
PDF output:
![grafik](https://user-images.githubusercontent.com/7644776/161733675-15d564ab-2c9a-411f-8351-6f0ba5678bba.png)

DOCX output:
![grafik](https://user-images.githubusercontent.com/7644776/161733761-c0e6662c-92c9-4cd9-8e50-4d21283b8d42.png)

No cropping - though the overflow property for the data item PRODUCTNAME says 
![grafik](https://user-images.githubusercontent.com/7644776/161733935-5bf0c57c-bd71-41cf-a264-932bc5b8144f.png)

Looking at the source code of `AbstractWordXmlWriter.java` I would expect the cropping you describe - but I can't reproduce.

**Reply by reiner-killet** [reply]
the behaviour desribed in [https://github.com/eclipse/birt/issues/875] is not what we see in 4.8 and 4.9. There is no line break. It deletes 2 digits and the decimal sign:
![grafik](https://user-images.githubusercontent.com/95427074/161745249-f49a60ec-0a06-4673-bb47-c6d1afa98ffe.png)
In my sample template the data type is "Float", tag number format is "Fixed" with grouping characters and 2 decimal places. The cell ist right-aligned.

**Reply by hvbtup** [reply]
I don't say this is related to #875. I only used the existing report from #875 to test this.

**Reply by hvbtup** [reply]
BTW you shouldn't use float (in the DB or in BIRT or anywhere) to store prices. It is complicated enough with decimal data and 19% MWSt in Germany :smiley:, so you don't want additional conversion errors.

**Comment by reiner-killet** [comment]
Enclosed please find template & XML data.

Von: Henning von Bargen ***@***.***>
Gesendet: Dienstag, 5. April 2022 15:21
An: eclipse/birt ***@***.***>
Cc: Reiner Killet ***@***.***>; Mention ***@***.***>
Betreff: Re: [eclipse/birt] property overflow = Hidden leads to cut-off decimals with DOCX output (Discussion #916)


Please attach the rptdesign in a *.zip file.

**Reply by hvbtup** [reply]
You can drag a ZIP file directly to the editor control here...

**Reply by reiner-killet** [reply]
[Test.zip](https://github.com/eclipse/birt/files/8418459/Test.zip)

**Comment by hvbtup** [comment]
I think the problem here is that "380.000.000,00" (meaning 380 millions in German) is considered as _one word_. And the representation of this word in Arial 10pt is wider than the column width specified. In my test report from #875 the long text consists of several words where each one fits into the column. That explains why it cannot be reproduced with that report.

The behavior is similar to the PDF emitter if the `pdfRenderOption.wordBreak` option is not set (as can be seen with Run Report As ... PDF).

By setting aforementioned option, the PDF emitter can be configured to break a word if it is too long even if it does not contain spaces.

Maybe the DOCX emitter could use a similar option which would avoid calling the cropping in AbstractWordWriterXml.java if it's set, thus restoring the behavior before the commit https://github.com/eclipse/birt/commit/8c4f519544b443fedc3b2d89fdbd8020a7dc320d). Search for "crop" in the Java source file to see what I mean.

@reiner-killet Can you create a PR for this?

**Comment by reiner-killet** [comment]
The XML data source does not contain 380.000.000,00 
It contains 380000000.17
In my first screenshot you can see that it evem crops numbers although there is a lot of space in the cell.
Let us assume that property "Overflow" was added by 4.8 when opeing a template that was created in an older version.
Unfortunately, the default was "Hidden" and not "Visible".
That's the reason for our misery, correct? 
In this case we should simply replace all
   `<property name="overflow">auto</property>`     (new remplates)
   `<property name="overflow">hidden</property>` (old templates)
by
   `<property name="overflow">visible</property>`
in the entire rptDesign, maybe using Notepad++.
I guess this will cause an appropriate DOCX output, correct?

**Reply by hvbtup** [reply]
I think that your assumption about what's going on is wrong. The property has "always" been there, but the Word emitter just ignored it until the commit https://github.com/eclipse/birt/commit/8c4f519544b443fedc3b2d89fdbd8020a7dc320d in 2017.

As I said, the problem seems to occur only if single words (or numbers) are too long to fit into a single line (not considering the edge case display="inline").

It should suffice to modify only those items (label, data, text, dynamic text) of your reports where this might be a problem.
The tricky part is how to identify those, depending of the amount of your reports.

Unfortunately, I believe that your assumption about what to do in order to avoid the problem is wrong, too.
I bet you won't find the overflow property set explicitly everywhere. So you can't use this as a search condition.
Instead you would have to find items where the overflow is not specified, too.

But you can easily test this yourself, of course.

If I'm right, then Notepad++ won't work. You'd have to write a little program that "understands XML".

But as I said, only really long words or numbers seem to cause the problem. If we assume that cropping a very long word is just an annoyance to the reader but not "bad enough" to render the whole report useless, then you can concentrate on those columns where you output numbers. In that case it may be reasonable to check and edit the reports manually in the designer.

**Comment by hvbtup** [comment]
In #1013 I proposed a way of setting overflow to visible without changing each and every label and (dynamic) text item. It works by using styles.
